frame/safe-mode/src/lib.rs

```diff
@@ -393,25 +434,39 @@ impl<T: Config> Pallet<T> {
 		Ok(())
 	}
 
-	/// Logic for the [`crate::Pallet::release_reservation`] call.
+	/// Logic for the [`crate::Pallet::release_reservation`] and
+	/// [`crate::Pallet::force_release_reservation`] calls.
 	///
-	/// Errors if the safe-mode is active.
-	/// Does not check the origin.
-	fn do_release_reservation(account: T::AccountId, block: T::BlockNumber) -> DispatchResult {
-		ensure!(!Self::is_activated(), Error::<T>::IsActive);
-		let reserve = Reservations::<T>::take(&account, block).ok_or(Error::<T>::NoReservation)?;
+	/// Errors if the safe-mode is active with [`Error::IsActive`].
+	/// Errors if release is called too soon by anyone but [`Config::ForceReservationOrigin`] with
+	/// [`Error::CannotReleaseYet`]. Does not check the origin.
+	fn do_release(
+		who: Option<T::AccountId>,
+		account: T::AccountId,
+		block: T::BlockNumber,
+	) -> DispatchResult {
+		ensure!(!Self::is_active(), Error::<T>::IsActive);
+
+		let reserve = Reservations::<T>::get(&account, &block).ok_or(Error::<T>::NoReservation)?;
+
+		if who.is_some() {
+			let delay = T::ReleaseDelay::get().ok_or(Error::<T>::NotConfigured)?;
+			let now = <frame_system::Pallet<T>>::block_number();
+			ensure!(now > (block + delay), Error::<T>::CannotReleaseYet);
```

Comment:
Note: the `release_reservation` call allows anyone to release a deposit after a configured timeout (if `ReleaseDelay` is set to `Some(delay)`) , but this is based on the block the reservation was taken at, not the time the safe mode was deactivated. Thus when deactivated release may be called immediately if that delay time has elapsed while the mode is activated. We could go for a new storage item that keeps the last block safe mode was active at and base _all_ releases by any signed origin to be possible only after that delay based on the last activated time. IT is unclear to me if we can get a more nuanced release cycle per reservation that takes into account the last time safe mode was activated (or if that would even be something we want)

Reply:
Should use safe math here